File: Python/other/remove_outermost_parenthesis.py
def removeOuterParentheses(S):
    res, opened = [], 0
    for c in S:
        if c == '(' and opened > 0:
            res.append(c)
        if c == ')' and opened > 1:
            res.append(c)
        opened += 1 if c == '(' else -1
    return "".join(res)

str = input("Enter string: ")
result = removeOuterParentheses(str)
print(result)


"""
Time complexity: O(n)
Space complexity: O(n)
Sample Input 1:
Enter string: (()())(())
Output: ()()()

Sample Input 2:
Enter string: ()()
Output: ""

"""

File: C-Plus-Plus/other/remove_outermost_parenthesis.cpp
#include <bits/stdc++.h>
using namespace std;
//Main Begin
int main()
{
    string str, res;
    //Count variable take care
    //of number of count prsent
    int count = 0;
    //Input the String
    cout << "Enter String :" << endl;
    cin >> str;
    //For Each loop
    //traverse the entire str
    for (auto ch : str) {
        /*Check if '(' is present,
    	    then increment count variable
    	    by +1
    	    else decrement the count variable 
    	    by -1
    	    */
        if (ch == '(') {
            ++count;
            //if count >0 then
            //add ch to res String
            if (count != 1) {
                res += ch;
            }
        }
        else {
            --count;
            //if count >1 then
            //add ch to res String
            if (count != 0) {
                res += ch;
            }
        }
    }
    //Print the final output
    cout << "Output:" << res;
    return 0;
}

/* Sample Input Output 1.
Enter String : "(()())(())"
Output:  ()()()
Sample Input Output 1.
Enter String : "()()"
Output: ""
Complexities :
Time Complexity : O(n)
Space Complexity : O(n)
*/
